contracts/extensions/CoinMachine.sol

```diff
@@ -130,53 +157,82 @@ contract CoinMachine is ColonyExtension {
 
     targetPerPeriod = _targetPerPeriod;
     maxPerPeriod = _maxPerPeriod;
-    tokensToSell = _tokensToSell;
 
     activePrice = _startingPrice;
     activePeriod = getCurrentPeriod();
 
     emaIntake = wmul(targetPerPeriod, _startingPrice);
 
+    setPriceEvolution(getTokenBalance() > 0);
+
     emit ExtensionInitialised();
   }
 
+  /// @notice Set the address for an (optional) whitelist
+  /// @param _whitelist The address of the whitelist
+  function setWhitelist(address _whitelist) public onlyRoot notDeprecated {
+    whitelist = _whitelist;
+
+    emit WhitelistSet(_whitelist);
+  }
+
   /// @notice Purchase tokens from Coin Machine.
   /// @param _numTokens The number of tokens to purchase
   function buyTokens(uint256 _numTokens) public payable notDeprecated {
     updatePeriod();
 
-    uint256 numTokens = min(min(_numTokens, maxPerPeriod - activeSold), tokensToSell);
+    require(
+      whitelist == address(0x0) || Whitelist(whitelist).isApproved(msg.sender),
+      "coin-machine-unauthorised"
+    );
+
+    uint256 tokenBalance = getTokenBalance();
+    uint256 numTokens = min(min(_numTokens, maxPerPeriod - activeSold), tokenBalance);
     uint256 totalCost = wmul(numTokens, activePrice);
 
     activeIntake = add(activeIntake, totalCost);
     activeSold = add(activeSold, numTokens);
-    tokensToSell = sub(tokensToSell, numTokens);
 
     assert(activeSold <= maxPerPeriod);
 
+    // Check if we've sold out
+    if (numTokens >= tokenBalance) { setPriceEvolution(false); }
+
     if (purchaseToken == address(0x0)) {
       require(msg.value >= totalCost, "coin-machine-insufficient-funds");
-      if (msg.value > totalCost) { msg.sender.transfer(msg.value - totalCost); }
+      if (msg.value > totalCost) { msg.sender.transfer(msg.value - totalCost); } // Refund any balance
+      payable(address(colony)).transfer(totalCost);
     } else {
-      require(ERC20Extended(purchaseToken).transferFrom(msg.sender, address(this), totalCost), "coin-machine-transfer-failed");
+      require(ERC20(purchaseToken).transferFrom(msg.sender, address(colony), totalCost), "coin-machine-purchase-failed");
     }
 
-    colony.mintTokensFor(msg.sender, numTokens);
+    require(ERC20(token).transfer(msg.sender, numTokens), "coin-machine-transfer-failed");
 
-    emit TokensBought(msg.sender, _numTokens, totalCost);
+    emit TokensBought(msg.sender, numTokens, totalCost);
   }
 
   /// @notice Bring the token accounting current
   function updatePeriod() public {
     uint256 currentPeriod = getCurrentPeriod();
     uint256 initialActivePeriod = activePeriod;
 
+    if (initialActivePeriod >= currentPeriod) {
+      return;
+    }
+
+    // If we're out of tokens or deprecated, don't update
+    if (!evolvePrice) {
+      activePeriod = currentPeriod;
+
+      // Are we still sold out?
+      setPriceEvolution(getTokenBalance() > 0 && !deprecated);
```

Comment:
Both `initialise` and `finishUpgrade` can be called while deprecated, but they don't have `&& !deprecated` in the call to `setPriceEvolution`, which I think they should?

Reply:
I thought about that, ultimately I didn't think that either would ever be called while deprecated. But I can add it if you like.